docs/architecture/adr-108-e2e-abci++.md

```diff
@@ -0,0 +1,296 @@
+The resulting code is inside the following directories: 
+- `test/e2e/pkg/grammar/clean-start/lexer`,
+- `test/e2e/pkg/grammar/clean-start/parser`,
+- `test/e2e/pkg/grammar/clean-start/sppf`,
+- `test/e2e/pkg/grammar/clean-start/token`,
+- `test/e2e/pkg/grammar/recovery/lexer`,
+- `test/e2e/pkg/grammar/recovery/parser`,
+- `test/e2e/pkg/grammar/recoveryt/sppf`,
+- `test/e2e/pkg/grammar/recovery/token`.
+
+Apart from this auto-generated code, we implemented `GrammarChecker` abstraction
+which knows how to use the generated parsers and lexers to verify whether a
+specific execution (list of ABCI++ calls logged by the Application while the
+testnet was running) respects the ABCI++ grammar. The implementation and tests 
+for it are inside `test/e2e/pkg/grammar/checker.go` and 
+`test/e2e/pkg/grammar/checker_test.go`, respectively. 
+
+How the `GrammarChecker` works is demonstrated with the test `TestABCIGrammar`
+implemented in `test/e2e/tests/abci_test.go` file. 
+
+```go
+func TestABCIGrammar(t *testing.T) {
+	checker := grammar.NewGrammarChecker(grammar.DefaultConfig())
+	testNode(t, func(t *testing.T, node e2e.Node) {
+		if !node.Testnet.ABCITestsEnabled {
+			return
+		}
+		reqs, err := fetchABCIRequests(t, node.Name)
+		if err != nil {
+			t.Error(fmt.Errorf("Collecting of ABCI requests failed: %w", err))
+		}
+		for i, r := range reqs {
+			isCleanStart := i == 0
+			_, err := checker.Verify(r, isCleanStart)
+			if err != nil {
+				t.Error(fmt.Errorf("ABCI grammar verification failed: %w", err))
+			}
+		}
+	})
+}
+```
+
+Specifically, the test first creates a `GrammarChecker` object. Then for each node in the testnet, it collects all requests 
+logged by this node. Remember here that `fetchABCIRequests()` returns an array of slices(`[]*abci.Request`) where the slice 
+with index 0 corresponds to the node's `CleanStart` execution, and each additional slice corresponds to the `Recovery` 
+execution after a specific crash. Each node must have one `CleanStart` execution and the same number of `Recovery` executions 
+as the number of crashes that happened on this node. If collecting was successful, the test checks whether each execution 
+respects the ABCI++ 
+grammar by calling `checker.Verify()` method. If `Verify` returns an error, the specific execution does not respect the 
+grammar, and the test will fail. 
+
+The tests are executed only if `ABCITestsEnabled` is set to `true`. This is done through the manifest file. Namely, if we 
+want to test whether CometBFT respects ABCI++ grammar, we would need to enable these tests by adding `abci_tests_enabled = 
+true` in the manifest file of a particular testnet (e.g. `networks/ci.toml`). This will automatically activate logging on the 
+application side. 
+
+The `Verify()` method is shown below. 
+```go
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	r := g.filterRequests(reqs)
```

Comment:
Here, you are using `reqs` and validating its contents just after. Shouldn't you move this line after 196? (so that you first validate `reqs` and then use it)

Reply:
.. or should line 194 use `r` rather than `reqs`?